tools/fdleak.py

```diff
@@ -0,0 +1,405 @@
+./fdleak -m 60
+        Set the minimum allowable survival time to 60 seconds
+"""
+
+parser = argparse.ArgumentParser(
+    description="Trace FD leak",
+    formatter_class=argparse.RawDescriptionHelpFormatter,
+    epilog=examples)
+parser.add_argument("-p", "--pid", default=-1, help="trace this PID only")
+parser.add_argument("--lsof", default=False, action="store_true", dest="lsof",
+        help="list the files opened by the monitor process")
+parser.add_argument("-i", "--interval", default=5, type=int,
+        help="interval (in seconds) to scan for outstanding allocations")
+parser.add_argument("-m", "--min_allow", default=30, type=int,
+        help="minimum allowable survival time")
+parser.add_argument("-D", "--debug", default=False, action="store_true",
+        help="print debug message, only for developer")
+parser.add_argument("--ebpf", action="store_true",
+        help=argparse.SUPPRESS)
+args = parser.parse_args()
+
+bpf_text = """
+#include <asm-generic/errno-base.h>
+#include <bcc/proto.h>
+#include <linux/fdtable.h>
+#include <linux/sched.h>
+#include <uapi/linux/ptrace.h>
+struct alloc_info_t {
+  u64 timestamp_ns;
+  int stack_id;
+  u32 tid;
+  int fd;
+  u32 pid;
+};
+
+struct combined_alloc_info_t {
+  //TODO Determine whether the same stack has different survival times
+  u64 last_longest_survivor;
+  int number_of_allocs;
+};
+
+BPF_STACK_TRACE(user_stack, 102400);
+BPF_HASH(pidfd_allocs_hash, u64, struct alloc_info_t, 102400);
+BPF_HASH(combined_allocs_hash, int, struct combined_alloc_info_t, 10240);
+//TODO
+BPF_HASH(pid_exit_hash, int, u8, 10240);
+
+// Programs that rely on the system `exit_files`are not considered
+// However, their fd is generally less than NR_OPEN_DEFAULT
+#define FD_MONITOR_TRIGGER(fd)        \
+  if (fd > 0 && fd < NR_OPEN_DEFAULT) \
+    return 0;
+
+static inline void update_statistics_add(int stack_id) {
+  struct combined_alloc_info_t *existing_cinfo;
+  struct combined_alloc_info_t cinfo = {0};
+  existing_cinfo = combined_allocs_hash.lookup(&stack_id);
+  if (existing_cinfo != NULL)
+    cinfo = *existing_cinfo;
+  cinfo.number_of_allocs += 1;
+  combined_allocs_hash.update(&stack_id, &cinfo);
+}
+
+static inline void clear_pid_exit_flag(){
+  int pid = bpf_get_current_pid_tgid()>>32;
+  u8 is_exit = 0;
+  pid_exit_hash.update(&pid, &is_exit);
+}
+
+static inline void set_pid_exit_flag(){
+  int pid = bpf_get_current_pid_tgid()>>32;
+  u8 is_exit = 1;
+  pid_exit_hash.update(&pid, &is_exit);
+}
+
+static inline int fd_alloc_return(struct pt_regs *ctx) {
+  int ret_fd = PT_REGS_RC(ctx);
+  struct alloc_info_t info = {0};
+  u64 pidfd = bpf_get_current_pid_tgid() & 0xffffffff00000000L;
+  u32 pid = bpf_get_current_pid_tgid()>>32;
+  u32 tid = (u32)(bpf_get_current_pid_tgid() & 0xffffffffL);
+  struct combined_alloc_info_t *cinfo;
+  //other errors are not the monitor focus
+  if(ret_fd < 0 && ret_fd != -EMFILE)
+    return 0;
+
+  pidfd += (u32)ret_fd;
```

Comment:
You use TGID + fd as the key, what if the below case happens?

TGID = 5000, fd ret = 10, pidfd = 5010
TGID = 5008, fd ret = 2, pidfd = 5010